### PR TITLE
Added query runner for Google Analytics Data API (GA4) and Google Search Console API 

### PR DESCRIPTION
**Pull request by vinodpandey**
## What type of PR is this? 

- [x] New Query Runner (Data Source) 

## Description
Added query runner for [Google Analytics Data API (GA4)](https://developers.google.com/analytics/devguides/reporting/data/v1 ) and [Google Search Console API](https://developers.google.com/webmaster-tools/about). For Google Analytics Data API (GA4), the query runner supports properties.runReport only (excluding pivots support)

## How is this tested?
- [x] Manually
- [ ] N/A

Google Analytics Data API (GA4)
- Create a new data source by specifying Name, GA4 Property ID and Authentication JSON file (the key file is same as we use in previous Google Analytics Query runner)
- After creating new data source, executed query with payload
```
{
  "dateRanges": [
    {
      "startDate": "yesterday",
      "endDate": "yesterday"
    }
  ],
  "dimensions": [
    {
      "name": "date"
    }
  ],
  "dimensionFilter": {
    "filter": {
      "fieldName": "pagePath",
      "stringFilter": {
        "matchType": "BEGINS_WITH",
        "value": "/"
      }
    }
  },
  "metrics": [
    {
      "name": "activeUsers"
    }
  ]
}
```
- it return data with date and activeUsers column

Google Search Console API
- Create a new data source by specifying Name, Site URL and Authentication JSON file (the key file is same as we use in Google Analytics Query runner)
- After creating new data source, executed query with payload
```
{
        "startDate": "2022-11-01",
        "endDate": "2022-11-01",
        "dimensions": ["query", "page"],
        "rowLimit": 10,
        "startRow": 0
    
}
```
- it returns data in correct format

### CODE REVIEW

**Comment by poborin** [comment]
I can confirm that the GA4 query runner works.

**Comment by shmokmt** [comment]
Is this PR in a state of pending review?

**Comment by mklinik** [comment]
I'd really love to have this. Can I help somehow to move it ahead?

**Comment by timing** [comment]
How can I help to make this PR happen? I'd love to have this! Also @vinodpandey, do you know how to make sure the checks succeed? It might be a first step to have this PR merged. Thanks for your work!

**Comment by vinodpandey** [comment]
Sorry @timing, it seems the checks are failing because of environment/tooling related reasons over which I have no control. The CircleCI pipeline complains about config.yml which I have not updated or removed in this pull request https://app.circleci.com/pipelines/github/getredash/redash. The netlify deploy is failing because of unsupported build image (https://app.netlify.com/sites/redash-preview/deploys/638b54730f1ebf00084f67c8). The image needs to be updated to Ubuntu 20.04 in netlify console.

**Comment by vinodpandey** [comment]
I have reverted the changes in offending files (Dockerfile and requirements_bundles.txt). The changes which I made, already existed in master branch and I had to make these changes in my fork for the application to work. If someone wants to test these changes in my fork they will have to update these files (Dockerfile and requirements_bundles.txt) first after code checkout. 

Below are the changes:

Dockerfile -> https://github.com/getredash/redash/commit/d468fcb68edf4c921046a97891b34c2f807827f5
requirements_bundles.txt -> https://github.com/getredash/redash/commit/d8c5409f4783641d7a867b2e360062a39a02c2ef

**Comment by vinodpandey** [comment]
Fixed merge conflicts for `redash/settings/__init__.py`. @guidopetri Please let me know if more changes are required. Thanks

**Comment by guidopetri** [comment]
Thanks! As Codecov bot has posted above, it seems we don't have any tests. Would you mind adding some?

**Comment by vinodpandey** [comment]
Thanks @guidopetri. Let me add test cases and update the PR.

File: redash/query_runner/google_analytics4.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+            data_type = "DATE"
+        elif h["name"] == "dateHour":
+            data_type = "DATETIME"
+        columns.append(
+            {
+                "name": h["name"],
+                "friendly_name": h["name"],
+                "type": types_conv.get(data_type, "string"),
+            }
+        )
```
Comment (guidopetri) [inline redash/query_runner/google_analytics4.py:85]: Since this is ~identical to L59-71, can we have a nested for loop instead to keep the code more DRY?

**Comment by guidopetri** [comment]
Thanks for adding those tests @vinodpandey ! I've got one small change request, but otherwise this LGTM.

**Comment by guidopetri** [comment]
@wlach @junnplus do you want to take a look as well?

**Comment by vinodpandey** [comment]
Thanks @guidopetri for code review and feedback. I have optimized the code and created a common method. I tried using  `zip_longest(response["dimensionHeaders"], response["metricHeaders"])` to create single for loop but that didn't work because of the response JSON structure of Google Analytics Data API. I am unsure of how to implement a nested for loop for this scenario.

To process new JSON structure (Documentation [link1](https://developers.google.com/analytics/devguides/reporting/data/v1/rest/v1beta/RunReportResponse) and [link2](https://developers.google.com/analytics/devguides/reporting/data/v1/rest/v1beta/Row ) ),  I am extracting column names first from dimensionHeaders and then from metricHeaders. The order is important here because while extracting the column values, I have to process first dimensionValues and then metricValues and map those to respective headers based on the position of the item in list.

File: redash/query_runner/google_analytics4.py (hunk 1 above)
Comment (vinodpandey) [reply redash/query_runner/google_analytics4.py]: Added a common method to fix the duplicate code.

**Review by guidopetri — APPROVED** [review]

**Comment by guidopetri** [comment]
Merged! Thanks again for your contribution :)

**Pull request merged**